python/ray/data/_internal/table_block.py

```diff
@@ -46,6 +48,20 @@ def add(self, item: Union[dict, TableRow, np.ndarray]) -> None:
                 "Returned elements of an TableBlock must be of type `dict`, "
                 "got {} (type {}).".format(item, type(item))
             )
+
+        item_columns_names = sorted(item.keys())
+        if self._sorted_columns_names:
+            # Check all added rows have same columns.
+            if item_columns_names != self._sorted_columns_names:
+                raise ValueError(
+                    "Current row has different columns compared to previous rows. "
+                    f"Columns of current row: {item_columns_names}, "
+                    f"Columns of previous rows: {self._sorted_columns_names}."
+                )
+        else:
+            # Initialize columns names with the first added row.
+            self._sorted_columns_names = item_columns_names
```

Comment:
Hmm why use a sorted list instead of a set? I feel like a set is what we want here, and doesn't require sorting or even copying the column names until the check fails and we want to display the diff (off the critical path):
```suggestion
        item_column_names = item.keys()
        if self._column_names is not None:
            # Check all added rows have same columns.
            if item_column_names != self._column_names:
                raise ValueError(
                    "Current row has different columns compared to previous rows. "
                    f"Columns of current row: {sorted(item_column_names)}, "
                    f"Columns of previous rows: {sorted(self._column_names)}."
                )
        else:
            # Initialize column names with the first added row.
            self._column_names = item_column_names
```

Reply:
@clarkzinzow - good point, updated.